### PR TITLE
4 - pull bill data

### PR DESCRIPTION
**Pull request by edhille**
- Moved to umbrella structure so data-fetch is it's own application
- Set up fetching for Bill data from GPO.gov
- Set up fetching for Member and Committee data from Propublica (NOTE: API key is in "secret" config files not committed, so ping me if you do not have a Propublica key you can use)

### CODE REVIEW

File: apps/action_data_fetcher/lib/action_data_fetcher/gpo/fetcher/worker.ex [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    # TODO: should this also be configured?
+    "https://www.gpo.gov/fdsys/bulkdata/BILLSTATUS/#{congress}/#{bill_type}/BILLSTATUS-#{congress}-#{bill_type}.zip"
+  end
+
+  defp write_zip_to_tmpdir(congress, bill_type, zip_data) do
+    date_time_str = String.replace(DateTime.to_string(DateTime.utc_now()), " ", "_")
+    path          = Path.join([@tmp_dir, Integer.to_string(congress), bill_type])
+    file_path     = Path.join([path, "#{date_time_str}.zip"])
+
+    case @file_module.mkdir_p(path) do
```
Comment (sodiumjoe) [inline apps/action_data_fetcher/lib/action_data_fetcher/gpo/fetcher/worker.ex:59]: i think you could unnest this a bit with `with`

File: apps/action_data_fetcher/lib/action_data_fetcher/gpo/fetcher/worker.ex [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  @sys_module Application.get_env(:action_data_fetcher, :gpo)[:sys_module] || System
+  @tmp_dir Application.get_env(:action_data_fetcher, :tmp_dir)
+
+  defmodule FetchState do
+    defstruct [:congress, :bill_type, :zip_data]
+  end
+
+  ## Worker API
+
+  def start_link(stuff) do
```
Comment (sodiumjoe) [inline apps/action_data_fetcher/lib/action_data_fetcher/gpo/fetcher/worker.ex]: `stuff`?

File: apps/action_data_fetcher/lib/action_data_fetcher/gpo/fetcher/worker.ex [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    generate_url(congress, type) 
+    |> @http.get([timeout: 90_000]) # TODO: Magic constant
+  end
+
+  defp generate_url(congress, bill_type) do
+    # TODO: should this also be configured?
+    "https://www.gpo.gov/fdsys/bulkdata/BILLSTATUS/#{congress}/#{bill_type}/BILLSTATUS-#{congress}-#{bill_type}.zip"
+  end
+
+  defp write_zip_to_tmpdir(congress, bill_type, zip_data) do
```
Comment (sodiumjoe) [inline apps/action_data_fetcher/lib/action_data_fetcher/gpo/fetcher/worker.ex:54]: it's not clear to me why the tmp file is necessary. Can we just use http://elixir-lang.org/getting-started/erlang-libraries.html#the-zip-and-zlib-modules to do this in memory? We're not going to rely on the file artifacts in any way, are we?

Comment (sodiumjoe) [reply apps/action_data_fetcher/lib/action_data_fetcher/gpo/fetcher/worker.ex:54]: not that this should block merging, could just be another @todo

File: apps/action_data_fetcher/lib/action_data_fetcher/gpo/parser/worker.ex [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  use GenServer
+  import SweetXml
+
+  ## Worker API
+
+  def start_link(stuff) do
+    GenServer.start_link(__MODULE__, stuff)
+  end
+  
+  def handle_call({:parse_bill, {:filepath, bill_data_xml_path}}, _from, state) do
```
Comment (sodiumjoe) [inline apps/action_data_fetcher/lib/action_data_fetcher/gpo/parser/worker.ex]: what `state` do these workers have?

File: apps/action_data_fetcher/lib/action_data_fetcher/gpo/server.ex [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      |> Enum.map(fn(task) ->
+        {:ok, path} = Task.await(task, @timeout)
+        results = list_files(path)
+          |> Enum.map(&queue_parse(&1))
+          |> Enum.map(fn(task) -> Task.await(task, @timeout) end)
+        File.rm_rf(path)
+        results
+      end)
+      |> List.flatten() # above creates list of result lists, so flatten it...
+      |> Enum.map(fn(response) -> response end)
```
Comment (sodiumjoe) [inline apps/action_data_fetcher/lib/action_data_fetcher/gpo/server.ex:42]: why is this necessary?

File: apps/action_data_fetcher/lib/action_data_fetcher/gpo/supervisor.ex [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+  defp poolboy_parser_spec do
+    poolboy_config = Application.get_env(:action_data_fetcher, :pools)[:gpo_parsers]
+
+    :poolboy.child_spec(:gpo_parsers, poolboy_config, [])
+  end
+
+  defp start_pool(poolboy_spec) do
+    case Supervisor.start_child(__MODULE__, poolboy_spec) do
+      {:pool_started, pool_pid} ->
```
Comment (sodiumjoe) [inline apps/action_data_fetcher/lib/action_data_fetcher/gpo/supervisor.ex:38]: what's the difference between `:pool_started` and `:ok`?

**Review by sodiumjoe — APPROVED** [review]
This is awesome work, ted! Everything looks pretty straightforward and readable. Thanks for being the main effort on the project!

File: apps/action_data_fetcher/lib/action_data_fetcher/gpo/fetcher/worker.ex (hunk 2 above)
Comment (edhille) [reply apps/action_data_fetcher/lib/action_data_fetcher/gpo/fetcher/worker.ex]: oof..forgot about that.  Will clean up

File: apps/action_data_fetcher/lib/action_data_fetcher/gpo/fetcher/worker.ex (hunk 3 above)
Comment (edhille) [reply apps/action_data_fetcher/lib/action_data_fetcher/gpo/fetcher/worker.ex:54]: hmm...not sure we want to unzip to memory and not sure what a zipped directory of files unzips to.  I'll check into it, but I think this is the route we want to take here.

File: apps/action_data_fetcher/lib/action_data_fetcher/gpo/parser/worker.ex (hunk 4 above)
Comment (edhille) [reply apps/action_data_fetcher/lib/action_data_fetcher/gpo/parser/worker.ex]: none. this is a questionable part of GenServers tbh. I put state to future-proof, but perhaps this is bad form?

File: apps/action_data_fetcher/lib/action_data_fetcher/gpo/server.ex (hunk 5 above)
Comment (edhille) [reply apps/action_data_fetcher/lib/action_data_fetcher/gpo/server.ex:42]: good question. I'll remove

File: apps/action_data_fetcher/lib/action_data_fetcher/gpo/supervisor.ex (hunk 6 above)
Comment (edhille) [reply apps/action_data_fetcher/lib/action_data_fetcher/gpo/supervisor.ex:38]: shoot...I cannot recall and I think this was a case where I was coding around responses. I'll double-check

File: apps/action_data_fetcher/lib/action_data_fetcher/gpo/fetcher/worker.ex (hunk 1 above)
Comment (edhille) [reply apps/action_data_fetcher/lib/action_data_fetcher/gpo/fetcher/worker.ex:59]: I think I need help doing that. I tried, but got confused on the fact that I have two different matches for each case.

Comment (sodiumjoe) [reply apps/action_data_fetcher/lib/action_data_fetcher/gpo/fetcher/worker.ex:59]: i think it would just be

```elixir
with
  :ok <- @file_module.mkdir_p(path)
  :ok <- @file_module.write(file_path, zip_data, [:binary, :write])
  do: {:ok, file_path}
```

I guess that doesn't keep the granular error type, but that doesn't seem super important to me.

File: apps/action_data_fetcher/lib/action_data_fetcher/gpo/parser/worker.ex (hunk 4 above)
Comment (sodiumjoe) [reply apps/action_data_fetcher/lib/action_data_fetcher/gpo/parser/worker.ex]: ¯\_(ツ)_/¯

File: apps/action_data_fetcher/lib/action_data_fetcher/gpo/fetcher/worker.ex (hunk 3 above)
Comment (sodiumjoe) [reply apps/action_data_fetcher/lib/action_data_fetcher/gpo/fetcher/worker.ex:54]: > not sure we want to unzip to memory

are you worried about the size?

> not sure what a zipped directory of files unzips to

good point, me neither

**Review by sodiumjoe — APPROVED** [review]
